Mise en place des analytics utilisateur et de la catégorie 'Dernières applications utilisées'

mamakh opened the pull request:
## Description

Design : PR #3 

Mise en place de la fonctionnalité 'Dernières applications utilisées'.
Nous avons mis en place une fonctionnalité d'analytics afin de stocker l'application à chaque navigation d'un utilisateur.
Nous pouvons stocker aussi le tenant sélectionné par l'utilisateur.
Cette API pourra être utilisé pour stocker à l'avenir d'autres informations.

## Type de changement

-  Code et/ou Configuration : Nouvelle fonctionnalité

## Tests

Navigation entre applications et vérification de la catégorie "Dernières utilisées".
Vérification des données stockées dans la base Mongo

## Migration

Aucune. Il n'y a pas d'analytics jusqu'à la mise en place avec cette PR.
La structure suivante apparaîtra dans les informations utilisateur : 
"analytics" : {
    "applications" : [
    {
        "applicationId" : "ARCHIVE_APP",
        "lastAccess" : "2020-01-09T00:00:00.000+01:00",
        "countAccess" : 5
    },
    {
        "applicationId" : "PORTAL_APP",
        "lastAccess" : "2020-01-09T00:00:00.000+01:00",
        "countAccess" : 11
    }
    ]
}


## Checklist

 [x] Mon code suit le style de code de ce projet.
 [x] J'ai commenté mon code, en particulier dans les classes et les méthodes difficile à comprendre.
 [x] J'ai rajouté les tests unitaires vérifiant mes fonctionnalités.
 [x] J'ai rajouté les tests de non régression vérifiant mes fonctionnalités.
 [x] Les tests unitaires nouveaux et existants passent avec succès localement.

## Contributeur

Xelians

fadil-zemmari-xelians committed:
[US TRTL-170] : save last used applications

[US TRTL-170] : code review - rename applicationAnalytics class

[US TRTL-170] : tag application us last used only at the first visit

[US TRTL-170] : code review

fadil-zemmari-xelians committed:
[US TRTL-172] : Get last used app back-end

[US TRTL-172] : Get last used app front-end
[US TRTL-172] : Fix import

[US TRTL-172] : Get last used apps update
[US TRTL-172]: Fix after conflict resolve
[US TRTL-172]: Fix last used app analytics loading

[US TRTL-172]: Fix duplicate when searching an application

[US TRTL-172]: Code review improvement

Commit:
[US TRTL-157] component SelectLanguage & VitamTranslateService

[US TRTL-157] MultiLoadHttpLoader & translate templates

[US TRTL-157]add gulpfile to migrate from i18n to ngx-translate

[US TRTL-157] save/update language & sessionStorage

[US TRTL-157] sessionStorage & add ROLE_UPDATE_ME_USERS for portal & delete gulpfile.js

[US TRTL-157] clean i18n for portal in conf files

[US TRTL-157] review & fix ngx-translate version

Commit:
[US TRTL-427] add httpLoaderChildFactory to use local assets/i18n

fadil-zemmari-xelians committed:
[TRTL-484] Fix ngx translate with lazy modules

fadil-zemmari-xelians committed:
[US TRTL-473] Hide selection language button

fadil-zemmari-xelians committed:
[TECH] Fix ngx translate unit tests

Commit:
[TECH] Update categories for deployment & local configurations

Commit:
Merge remote-tracking branch 'origin/feature/turtle/translate' into dlab-develop

Mocteezuma commented on ui/ui-frontend-common/src/app/modules/api/user-api.service.ts:
```diff
@@ -0,0 +1,57 @@
+    this.apiUrl = baseUrl + '/users';
+  }
+
+  public create(data: { applicationId?: string }): Observable<User> {
+    return this.http.post<User>(this.apiUrl + '/analytics', data);
+  }
```
Ca ne devrait pas plutôt être un PATCH sur user ?

Ne serait-ce que le nom de la fonction ? create porte a confusion pour moi. un update ou 'setLastApplication' me semble plus pertinant ?
En particulier, cela réduit je trouve la compréhension du resolver.

De plus, si j'ai bien compris le fonctionnement, ca fait soit du create soit du update sur un applicationAnalytics.
Et dans tout les cas un update du sous objet Analytics visé ici (et pas un create).

La remarque en bien entendue également valable pour la partie controller correspondante.

mamakh replied:
Nous avions décidé pour la partie analytics d'avoir une API dédiée et ne pas utiliser le patch user.
La partie analytique peut évoluer pour englober toute la gestion analytics du système c'est pour cela aussi que nous avons mis en place l'API dédiée.
Du coup sans réellement savoir comment c'est gérer au niveau du backend, on indique juste qu'il y a une nouvelle donnée d'analytics d'où le POST

Mocteezuma commented on api/api-iam/iam-internal/src/main/java/fr/gouv/vitamui/iam/internal/server/user/domain/ApplicationAnalytics.java:
```diff
@@ -0,0 +1,30 @@
+
+    public void tagAsLastUsed() {
+        this.lastAccess = OffsetDateTime.now();
+        this.accessCounter++;
```
Que se passe t-il lorsque accessCounter atteint Integer.MAX_VALUE ( +1 ) ?

Bien que la valeur soit très grande, elle est loin d'être inatégnable pour des utilisateurs utilisant régulièrement plusieurs applications.

delphine-coursier replied:
Integer.MAX_VALUE = 2147483647
Si un utilisateur accède 1000 fois par jour à la même application, cela 365 jours de l'année, pendant 100 ans, on atteindra 1000 x 365 x 100 = 36500000. Il lui restera encore 2110983647 accès avant d'atteindre le max. Je pense qu'on est large :joy:

Mocteezuma replied:
ah oui tiens, j'avais en tête un nombre plus petit. En effet

Mocteezuma reviewed (commented):
Une première review pour avancer sur le sujet en attendant la présentation de jeudi.

Globalement la manière d'implémenter me gène un peu.
Même si c'est du lien 1-1, avoir Analytics en sous-objet de User est particulier:
* Bien que ca soit des données de user se sont des données qui ne sont pas modifiable par l'utilisateur et qui ne renseignent rien de l'utilisateur (=l'objet metier me semble être different).
* Du coup, on a des points d'API create sur user qui ont pour but de le patch (en http POST) la collection User.
* Ca oblige aussi a patcher user très régulièrement alors que le User en particulier n'est pas vraiment mis à jour.

Je vais pas demander un changement complet de la logique, mais avez-vous évaluer cette piste ? Si oui pourquoi ne pas l'avoir retenue ?
Aussi, voyez-vous potentiellement cet objet analytics évoluer dans le temps ? Avec potentiellement des données sur les applications ou d'autres données globales ?

Commit:
[TECH] Remove tar.gz assembly

Commit:
[TECH] PR review fix for moment-mini which was unused

CouzinieAtCines commented on api/api-iam/iam-external/src/test/java/fr/gouv/vitamui/iam/external/server/rest/UserExternalControllerTest.java:
```diff
@@ -75,6 +76,15 @@ public void patchMe_thenOk() throws Exception {
 
+    @Test
+    public void patchAnalytics_thenOk() throws Exception {
+        LOGGER.debug("testPatchAnalytics");
```
Log pas vraiment utile ?

youcefxelians replied:
supprimé

CouzinieAtCines commented on api/api-iam/iam-external-client/src/test/java/fr/gouv/vitamui/iam/external/client/UserExternalRestClientTest.java:
```diff
@@ -0,0 +1,78 @@
+        verify(restTemplate).exchange(any(), any(), any(), eq(UserDto.class));
+    }
+
+    private UserDto buildUser() {
+        ApplicationAnalyticsDto applicationAnalytic = new ApplicationAnalyticsDto();
+        applicationAnalytic.setAccessCounter(85);
+        applicationAnalytic.setLastAccess(OffsetDateTime.now());
+        applicationAnalytic.setApplicationId("ACCOUNTS_APP");
+
+        AnalyticsDto analytics = new AnalyticsDto();
+        analytics.setApplications(List.of(applicationAnalytic));
+
+        UserDto user = new UserDto();
+        user.setId("78");
+        user.setEmail("test@user.fr");
+        user.setAnalytics(analytics);
+
+        return user;
+    }
```
Code identique utilisé dans plusieurs classes de test. Peut-être à refacto dans une classe utils accessible à tous les tests ?

pyXelians replied:
Concernant cette méthode, elle n'est utilisée que dans les classes de tests et avec des paramètres différents. De plus, elle est utilisée dans des modules différents. Je ne suis pas sûre que ça vaille le coup de la mutualiser avec une liste de paramètre.

CouzinieAtCines replied:
J'ai vu que c'est fait comme ca dans d'autres tests sur d'autres modules, d'où ma remarque.
Ca fait quand même toujours autant de lignes de code à maintenir et de tests à venir corriger le jour où il y a une évolution.
Après moi je considère que la meilleure ligne de code, c'est celle qui n'existe pas ^^

pyXelians replied:
Ok, j'ai fait le changement.

CouzinieAtCines commented on api/api-iam/iam-internal/src/main/java/fr/gouv/vitamui/iam/internal/server/user/service/UserInternalService.java:
```diff
@@ -871,6 +881,55 @@ public JsonNode findHistoryById(final String id) throws VitamClientException {
+    }
+
+    private void checkAnalyticsAllowedFields(final Map<String, Object> partialDto) {
+        Set<String> analyticsPatchAllowedFields = Set.of("applicationId");
```
Refacto "applicationId" au fichier de constantes CommonConstants.java ?

pyXelians replied:
Cette valeur est utilisée uniquement dans cette classe. Je ne pense pas qu'elle sera utilisée ailleurs. Si c'est le cas on  pourra toujours l'ajouter dans le CommonConstants à ce moment là.

CouzinieAtCines replied:
Je note 14 utilisations de cette chaine dans cette PR.

pyXelians replied:
Les 14 utilisations sont utilisés dans les classes de tests.  J'ai procédé au changement mais on se rend bien compte que toute l'application est concernée par cette pratique. J'ai changé l'applicationID, pour les autres il faudra faire une passe globale sur toute l'application pour utiliser des constantes mais ça dépasse le cadre de cette PR.

CouzinieAtCines commented on packaging/assembly-deployment.xml:
```diff
@@ -2,7 +2,7 @@
     <id>deployment</id>
     <formats>
-        <format>tar.gz</format>
+        <format>zip</format>
```
C'est quoi le rapport avec la choucroute ? 😅

pyXelians replied:
Il n'y a aucun rapport. C'est une modification que nous avions initialement faite dans une autre PR. Nous avons dû la reporter ici également  pour que le job sur Jenkins fonctionne, .

pyXelians committed:
[TECH] : Delete logger debug in test

pyXelians committed:
[TECH] Fix review - Create applicationIDConstant + userBuilder to test

pyXelians committed:
Merge branch 'feature/design/2' into feature/design/3

pyXelians committed:
[Tech] Fix package-lock

pyXelians committed:
[TECH] FIX merge

mamakh committed:
Choix du tenant par défaut & Sélection du tenant ou de l'organisation pour les apps multi-tenants et multi-organisations (#71)

* [US TRTL-172] : Get last used app back-end

[US TRTL-172] : Get last used app front-end
[US TRTL-172] : Fix import

[US TRTL-172] : Get last used apps update
[US TRTL-172]: Fix after conflict resolve
[US TRTL-172]: Fix last used app analytics loading

[US TRTL-172]: Fix duplicate when searching an application

[US TRTL-172]: Code review improvement

* [US TRTL-170] : save last used applications

[US TRTL-170] : code review - rename applicationAnalytics class

[US TRTL-170] : tag application us last used only at the first visit

[US TRTL-170] : code review

* [US TRTL-171] Persist last tenant identifier analytics

* [US TRTL-216] Add select tenant dialog on first user auth

[US TRTL-216] Fix last tenant identifier synchronization

[US TRTL-216] Improvement & add tenant by app filter

[US TRTL-216] Code improvement
[US TRTL-216] Add hierarchy app new portal components

[US TRTL-216] Fix select tenant component css
[US TRTL-169] Add select tenant inside app engine
[TECH] Add lastly used application translation

[US TRTL-169] Fix selected tenant display in app engine

[US TRTL-169] Code & architecture improvement

[US TRTL-169] Add translation & code review

[US-TRTL-216] Revert header features from identity

* [US TRTL-408] Add unknown app icon


[FIX TRTL-423] Fix tenant selection


[US TRTL-425] Add customer selection in new header


[TRTL-425] Code review

* [TECH] Apply Translate modifications in selecttenant & selectcustomer

* [TECH] Remove dev files for VITAM (these files are in .gitignore)

* [TECH] Fix TU

* [TECH] Delete ngOnDestroy not used

Co-authored-by: Fadil Zemmari <fadil.zemmari@xelians.fr>
Co-authored-by: Delphine <delphine.coursier@teamdlab.com>
Co-authored-by: Makhtar DIAGNE <makhtar.diagne@teamdlab.com>
Co-authored-by: pybelecalo <pierre-yves.belecalo@xelians.fr>

pyXelians committed:
Merge branch 'feature/design/2' into feature/design/3
